networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/BeaconChainMethods.java

```diff
@@ -145,15 +164,71 @@ public static BeaconChainMethods create(
-        spec.getGenesisSchemaDefinitions().getSignedBeaconBlockSchema(),
-        true,
-        beaconBlocksByRootHandler,
-        peerLookup);
+    final SignedBeaconBlockSchema phase0BlockSchema =
+        spec.forMilestone(SpecMilestone.PHASE0).getSchemaDefinitions().getSignedBeaconBlockSchema();
+
+    final BeaconBlocksByRootRequestMessageSchema requestType =
+        BeaconBlocksByRootRequestMessage.SSZ_SCHEMA;
+    final boolean expectResponseToRequest = true;
+
+    final SingleProtocolEth2RpcMethod<BeaconBlocksByRootRequestMessage, SignedBeaconBlock>
+        v1Method =
+            new SingleProtocolEth2RpcMethod<>(
+                asyncRunner,
+                BEACON_BLOCKS_BY_ROOT,
+                1,
+                rpcEncoding,
+                requestType,
+                expectResponseToRequest,
+                encoding ->
+                    RpcResponseDecoder.createContextFreeDecoder(encoding, phase0BlockSchema),
+                beaconBlocksByRootHandler,
+                peerLookup);
+
+    if (spec.getForkSchedule().getSupportedMilestones().contains(SpecMilestone.ALTAIR)) {
+      final ResponseSchemaSupplier<Bytes4, SignedBeaconBlock> v2SchemaSupplier =
+          createForkAwareSchemaSupplier(
+              spec, recentChainData, SchemaDefinitions::getSignedBeaconBlockSchema);
+      final SingleProtocolEth2RpcMethod<BeaconBlocksByRootRequestMessage, SignedBeaconBlock>
+          v2Method =
+              new SingleProtocolEth2RpcMethod<>(
+                  asyncRunner,
+                  BEACON_BLOCKS_BY_ROOT,
+                  2,
+                  rpcEncoding,
+                  requestType,
+                  expectResponseToRequest,
+                  encoding -> RpcResponseDecoder.createForkAwareDecoder(encoding, v2SchemaSupplier),
+                  beaconBlocksByRootHandler,
+                  peerLookup);
+
+      return VersionedEth2RpcMethod.create(
+          rpcEncoding, requestType, expectResponseToRequest, List.of(v2Method, v1Method));
+    } else {
+      return v1Method;
+    }
+  }
+
+  private static <T extends SszData>
+      ResponseSchemaSupplier<Bytes4, T> createForkAwareSchemaSupplier(
+          final Spec spec,
+          final RecentChainData recentChainData,
+          final Function<SchemaDefinitions, SszSchema<T>> getSchemaFromDefinitions) {
+    final Map<Bytes4, SszSchema<T>> cachedResults = new ConcurrentHashMap<>();
+    return (forkDigest) -> {
+      final SszSchema<T> cachedSchema = cachedResults.get(forkDigest);
+      if (cachedSchema != null) {
+        return Optional.of(cachedSchema);
+      }
+      final Optional<SszSchema<T>> schema =
+          recentChainData
+              .getMilestoneByForkDigest(forkDigest)
+              .map(spec::forMilestone)
+              .map(SpecVersion::getSchemaDefinitions)
+              .map(getSchemaFromDefinitions);
+      schema.ifPresent(s -> cachedResults.putIfAbsent(forkDigest, s));
+      return schema;
```

Comment:
Not entirely sure this is better but you can perform this operation atomically with `computeIfAbsent` because it treats `null` return values as meaning "don't insert anything".

```suggestion
      return Optional.ofNullable(
          cachedResults.computeIfAbsent(
              forkDigest,
              digest ->
                  recentChainData
                      .getMilestoneByForkDigest(digest)
                      .map(spec::forMilestone)
                      .map(SpecVersion::getSchemaDefinitions)
                      .map(getSchemaFromDefinitions)
                      .orElse(null)));
```

Reply:
Cutting this in the follow-up PR, so I'll just leave it for now